### DISCUSSION TITLE
Up to date player status / player data

### DISCUSSION THREAD

**Discussion by aaaclones** [kyleskom/NBA-Machine-Learning-Sports-Betting#172] [General]
There's no match up for player data and there's no up to date player status , This has to cause a big skew on the O/U and the E.v  , 
without key players the whole entire data batch might as well can be thrown out , for example Jokic was out and it totally skewed the ev and totals ... Is it possible to somehow manage to get the status of players and their data in real time ? this most certainly will create a better accuracy

**Comment by JP23000** [comment]
Yes it will.  This is one of the changes I have been trying to make but all of this is way outside the scope of my abilities.  The OU and ML would be so much more accurate if the dataset was based off of player stats and then checking the lineup each day for what players are out and then building the simulations from there.

**Comment by aaaclones** [comment]
There has to be some sort of framework or some sort of approach for this ..
The nba standard deviation for each team insist of 12 players but of the 12
we can safely assume it's normally 7 players on a regular basis , I am sure
if we can somehow get some how get lineup updates we can get it this to
 incredible accuracy levels , for example some players slow the game down
tremendously in the line up and some slow it down this most certainly
weighs heavily on O/U since more possessions will create a linear effect on
more pts , I am not familiar with how to code it into what's in the model
already , but the deviation on the day to day accuracy level definitely
needs improvement

**Comment by kyleskom** [comment]
Yes player data is the next crucial part of this that would take this to the next level. Its just not a simple problem and would take a fair amount of time which is hard to come by.

**Reply by aaaclones** [reply]
Kyle can you come up with some sort of framework on how to architect this I do have some ideas .. with player data ran with the team data I can't  help but to imagine what the accuracy level will be .. and can I ask why isn't the sheet updated daily?

**Reply by kyleskom** [reply]
What do you mean framework for that? The data isn't updated daily because ML forms general predictions on the data. The differences in daily updates would be so small that it wouldn't make any difference.

**Comment by aaaclones** [comment]
How would you go about adding player stats to combine it with the team
stats  ? Thanks for the clarification on the daily updates

**Reply by kyleskom** [reply]
You would need to get all player data for every day and also get the say starting lineups for every game. Then combine the player data with the players from the starting lineups for those days.

**Comment by BGB2k11** [comment]
I was thinking on helping on this whenever i have free time.
This can be a major feature and for real. this would make this model so huge .

Will keep you posted whenever ive found something

**Reply by BGB2k11** [reply]
This should be the link we need to fetch i presume ?
https://www.nba.com/stats/players/traditional

**Reply by mike-bostedt** [reply]
No we need each player on the roster for each game. So the model can correlate the wins to players. 
Does that make sence?

**Reply by kyleskom** [reply]
That is also needed. We need to combine those player stats with the players from the game.

**Reply by mike-bostedt** [reply]
You want seasonal avg stats from https://www.nba.com/stats/players/traditional to go along with the player data?
Or you want each players performance in each past game?

**Reply by mike-bostedt** [reply]
Actually Kyle you made it pretty clear here:
"You would need to get all player data for every day and also get the say starting lineups for every game. Then combine the player data with the players from the starting lineups for those days."

**Reply by kyleskom** [reply]
This is another good question. This is kinda why the player data aspect to all this is very tricky. Ill have to keep thinking on this.

**Reply by mike-bostedt** [reply]
What about adding a playerID to game data if minutes played is > 1min?

**Reply by nealmick** [reply]
@mike-bostedt That solution has worked for me, however finding perfect data with every player accurately marked for the past 10 years or so can be difficult.

https://github.com/nealmick/Merge-Data

Its not just a problem of historical data, you then also need current data.  You don't know who is going to play before a game... I have handled this by maintaining rosters of playerID by teamID, then gathering injury data, and allow users to decide to remove individual players based on doctors injury reports.  I will post two main API's i have used, they are rate limited, but i have crawled pretty much every endpoint listed and have saved all the data in pickle object files.

The main API used in my project is completely free, but will shut down for like 30 seconds if you request over the 1request/second rate limit.
https://www.balldontlie.io/home.html#introduction

Mostly paid API, I use for injury data and some other things, i have subscription with 2k request/day.
https://rapidapi.com/tank01/api/tank01-fantasy-stats/

**Reply by jbolter** [reply]
I would love to help solve this problem. I'm new to ML but have plenty of CS experience.  I was looking at this data set to solve this problem.  Seems like it will work but would love some feedback from the experts.  

https://www.kaggle.com/datasets/wyattowalsh/basketball
https://nba-db.readthedocs.io/

**Reply by nealmick** [reply]
I have found most dataset's to mark games with some sort of ID.  This might be originating from ESPN, NBA.com, or just completely made up.  From my experience games can be correlated across datasets using the game date, and the teams playing.  Sometimes i have used team ABV as-well but often data source abbreviate teams differently.  Players often time's can be correlated over multiple datasets using their name, so long as its spelled and formatted the same.

**Reply by kyleskom** [reply]
We would need to get the data from stats.nba.com and create our own. Creating the dataset isn't the hard part, its combining it to games.

**Comment by aaaclones** [comment]
past games is more important

**Comment by aaaclones** [comment]
player data is crucial , due to a lot of key players getting traded and key
players sitting out , that would skew the team data drastically .. I can't
see how this model can be efficient and variance will be all over the place
.. certainly the 8th guy on the bench wont do much to the team data but
what if it was a top 3 player on the team it will most certainly disrupt
the data set since the model trains itself based on what you trained it to
do just for team stats

**Reply by mike-bostedt** [reply]
I think everyone is in agreement that player data is the next step. We just need to work on implementation.

**Reply by kyleskom** [reply]
Yes

**Comment by aaaclones** [comment]
another reason i disagree with not updating the model regularly , Kyle you
mentioned wont do much or very little when you update the team data daily
but players are in and out so frequently .. the team data will most
certainly be effected

**Comment by aaaclones** [comment]
not sure how to go about it ? +/- for the player minutes a long with all
the standard data  and than compare and contrast to the team data

**Reply by mike-bostedt** [reply]
I don't understand this, can you elaborate?

**Comment by aaaclones** [comment]
you take each player on the current roster ... each player on the team has
a +/- effect on when they are on and off the floor , So for example
embid or jocik on and off the floor the +/- will
drastically definitely create some sort of ripple effect into the team data
.. this can allow you can take into account how much each player effects
the whole entire team data stats , some players do very to nothing but some
players will create a whole entire skew

**Comment by aaaclones** [comment]
I Dont know how to resolve this that's why i asked Kyle , I just can see
how some days when the standard deviation is normal the predictions are far
more accurate than other days were the lineup has been shuffled and
reshuffled

**Comment by kyleskom** [comment]
I think we would want players averages over the entire season like how the team data is. This way we get the best general model. When we grab player data for a prediction, the only way we can get a players predictions is going on season average. So I think player averages over the entire season would be the way to go. There are tons of stats to like +/- which would def be added as well. There's also a ton of advanced custom stats as well.

**Comment by kyleskom** [comment]
Getting the player data isn't the challenge here. Its getting players for all historical games and combining the player data with those games.

**Reply by mike-bostedt** [reply]
Where was the historical game summary data scraped from? I could start there to see if I can get the game data a level deeper.

**Reply by kyleskom** [reply]
It uses team data from every day and is combined with odds data. Team data comes from stats.nba.com

**Comment by aaaclones** [comment]
Perhaps you put a shorter historical reference ... You did 10 years for
team stats which i believe , maybe you do just the past 3 years or however
you want to implement it

**Reply by kyleskom** [reply]
The amount of years isn't the hard part its the implementation to do it at all. Once we automate 1 year many years is easy

**Reply by aaaclones** [reply]
player avg . for the season  x player vs. player 2 matchup historically x ( +/- ) with the current roster is this a tall order ?

**Reply by kyleskom** [reply]
Matchups like player v player offense vs defense? That would be very difficult. This needs to be more like, starting 5 players individual season averages for both teams.

**Comment by Marcfeitosa** [comment]
What do you think about applying sentiment analysis from Twitter? 
    Em sexta-feira, 7 de abril de 2023 às 16:00:51 BRT, Kyle Skompinski ***@***.***> escreveu:  
 
 


It uses team data from every day and is combined with odds data. Team data comes from stats.nba.com

**Reply by kyleskom** [reply]
Little to none added benefit.

**Comment by aaaclones** [comment]
This week has shown  you really why player data is so important the team
data has been tremendously skewed,  the confidence level on what the model
predictions are really so far off .. the neural network not so much as the
xgb , the O/U on what the delta has been has been has been scary and
misleading I hope people have taken this into account

**Reply by szocpaul** [reply]
Yes, I can confirm, this week was really bad compared to the last week, I was very profitable before with a ~70-75% win rate. I don't know what happened, maybe because of a lot of injuries.

<img width="845" alt="stake_bankroll" src="https://user-images.githubusercontent.com/10244579/230711721-46660587-b276-48ce-b084-09d05ac154c3.PNG">

**Reply by deperto** [reply]
Can you send the link to the bets placed and this graphic pls 😀🙏🏻

**Reply by Fastlane86** [reply]
It’s the end of the regular NBA season. A lot of teams that are already in the playoffs are resting their best players. Right now the B and C roaster of each team is playing. Unless it’s a team fighting for a seat at the play off table. They are still playing their A and B roaster. 

Anyways. What if we were to simplify player stats to scale of 1-100. Then give that value to a larger scale for the entire team that day. So if the starting line up is all roaster A. The team value would be higher in the 80s/90s or while a team playing all roaster C players would have a lower overall value. I know this is how lineups.com values teams and players.

**Reply by szocpaul** [reply]
> Can you send the link to the bets placed and this graphic pls 😀🙏🏻

https://app.bet-analytix.com/

**Reply by deperto** [reply]
Thank you, but i was referring to the actual link to see the bets ML and O/U, just to see the bets and strategy.

**Reply by szocpaul** [reply]
> Thank you, but i was referring to the actual link to see the bets ML and O/U, just to see the bets and strategy.

https://app.bet-analytix.com/bankroll/935156

**Comment by aaaclones** [comment]
@Neal did you give this a run yet ? after you merged the data ? i am going
through your fork but maybe you can save me some time and give me a yes/no
answer  , respectfully thank you

**Reply by nealmick** [reply]
No, I only built it to merge the data in the csv file and correlate games in my dataset to team entries on a particular day in kyle's dataset....  No future data, no model integration or adding the features, or anything like that.... The future data is the main difficult part... particularly injuries and constantly changing team rosters.

**Comment by aaaclones** [comment]
what if you ran the merged data before game time when more information has
been revealed about the roster ? What would that look like ?

**Reply by nealmick** [reply]
No future data, no model integration.... it just puts data in a csv :/

**Reply by nealmick** [reply]
xlsx really ha

**Comment by deperto** [comment]
Thank you 🙏🏻 


Sent from Yahoo Mail for iPhone


On Saturday, April 8, 2023, 19:43, szocpaul ***@***.***> wrote:





Thank you, but i was referring to the actual link to see the bets ML and O/U, just to see the bets and strategy.


https://app.bet-analytix.com/bankroll/935156

**Comment by aaaclones** [comment]
What data input will determine the 1-100 ? And some players have very
little to no data values , it's not just the bettors that are off the odds
makers are off the mark as well some of the lines are just so off ..  Let
say we are using the current model for the sake of argument how will you be
able to take 4-6 players that are on the floor  now that you have zero to
none information on to account for making your decision . I dont think you
can .. I think you have to take out the key players you made for the team
data and than determine how much it will effect what you been doing up to
current state..  The O/U line is the most determining factor on how good
your model truly is in my opinion and lately the xgb boost has been so far
off due to the data integrity since the lineup has been changed so
frequently

**Reply by kyleskom** [reply]
Betting the end of the season is pretty silly given the circumstances of teams tanking / resting players.
